src/StateMaps/PDEConstrainedFunctionals.jl

```diff
@@ -428,3 +428,342 @@ function Base.show(io::IO,::MIME"text/plain",f::AbstractPDEConstrainedFunctional
   print(io,"$(nameof(typeof(f))):
     num_constraints: $N")
 end
+
+############## Zygote Compat ##############
+"""
+    CustomPDEConstrainedFunctionals{N,A} <:  AbstractPDEConstrainedFunctionals{N}
+
+A version of `PDEConstrainedFunctionals` that allows for an arbitrary mapping
+`φ_to_jc` that is used to compute the objective and constraints given the primal variable.
+
+Under the hood, we use Zygote to compute the Jacobian of this mapping with the rrules defined
+throughout GridapTopOpt.
+
+# Parameters
+
+- `φ_to_jc`: A function that defines the mapping from the primal variable `φ` to the
+  objective and constraints. This should accept AbstractVector `φ` (the free values
+  of `φh`) and output a scalar. In side this function, you should compute your states
+  and evaluate your objectives and constraints that should be written as `GridapTopOpt.StateParamMap`.
+  For example,
+  ```julia
+  using GridapTopOpt: StateParamMap
+  ...
+  J = StateParamMap(j,state_map)
+  C = StateParamMap(c,state_map)
+  function φ_to_jc(φ)
+    u = state_map(φ)
+    [J(u,φ),C(u,φ)^2]
+  end
+  pcfs = CustomPDEConstrainedFunctionals(φ_to_jc,state_map,φh)
+  ```
+- `analytic_dJ`: Either a `Function` (if using analytic derivative) or nothing if using AD.
+- `analytic_dC`: A vector of `Nothing` or `Function` depending on whether using analytic derivatives or AD.
+- `state_map::A`: The state map for the problem. NOTE: this is a place holder for the optimiser
+  output and in theory you could use many different state maps inside `φ_to_jc`.
+
+
+!!! warning
+    The expected function for `analytic_dJ` and functions of `analytic_dC` are different to
+    usual. Here, you should define a function that takes an `AbstractVector` input corresponding to
+    the derivative and the primal variable dofs `φ` and assembles the derivative into the
+    `AbstractVector` input. For example,
+    ```julia
+    function analytic_dJ!(dJ,φ)
+      φh = FEFunction(V_φ,φ)
+      uh = get_state(state_map)
+      _dJ(q) = ∫(q*...)dΩ
+      Gridap.FESpaces.assemble_vector!(_dJ,dJ,V_φ)
+    end
+    ```
+    This functionality is subject to change in future!
+"""
+struct CustomPDEConstrainedFunctionals{N,A} <:  AbstractPDEConstrainedFunctionals{N}
+  φ_to_jc :: Function
+  analytic_dJ
+  analytic_dC
+  state_map :: A
+
+    @doc"""
+        CustomPDEConstrainedFunctionals(
+          φ_to_jc :: Function,
+          num_constraints,
+          state_map :: AbstractFEStateMap;
+          analytic_dJ = nothing,
+          analytic_dC = fill(nothing,num_constraints)
+        )
+
+    Create an instance of `CustomPDEConstrainedFunctionals`. Here,
+    `num_constraints` specifies the number of constraints.
+    """
+    function CustomPDEConstrainedFunctionals(
+      φ_to_jc :: Function,
+      num_constraints,

```

Comment:
Do you think there is a way to avoid manually specifying the number of constraints here and instead look at the length of the output of `φ_to_jc`? I am guessing you want to avoid having to evaluate the function in the constructor right?

Reply:
I don't think so, not without evaluating the the mapping. I don't like doing that as it is wasteful and balloons the time to first execution further.